File: Data Structures/Fenwick Tree BIT/BIT.java
import java.util.*;

public class BIT{
    private int size;
    private List<Long> tree;

    public BIT(List<Long> arr){
        this.size = arr.size();
        this.tree = new ArrayList<>(arr);

        for(int i = 1; i<= size; i++){
            int j = i + lsb(i);//pai de i
            if(j <= size)
                tree.set(j, tree.get(j)+tree.get(i));
            
        }
    }

    private int lsb(int n){
        return n & -n;
    }

    private Long querry(int idx){
        Long ans = 0;
        while(idx>0){
            ans+= tree.get(idx);
            idx-=lsb(idx);
        }
        return ans;
    }

    // delta = arr[idx] - value
    public incrementUpdate(int idx, Long delta){
        while(idx <= size){
            tree.set(idx, tree.get(idx) + delta);
            idx += lsb(idx);
        }
    }

    public Long querryRange(int init, int end){
        return querry(end) - querry(init-1);
    }

}

File: Data Structures/Fenwick Tree BIT/BIT.py
class BIT:
    def __init__(self, size: int) -> None:
        self.size = size
        self.tree = [0] * (size+1)
    
    def _lsb(n: int) -> int:
        return n&(-n)

    #indexado em 1
    def incrementUpdate(self, idx: int, delta: int) -> None:
        while(idx <= self.size):
            self.tree[idx]+=delta
            idx += self.lsb(idx)
        
    def querrys(self, idx: int) -> int:
        ans = 0
        while(idx > 0):
            ans += self.tree[idx]
            idx -= self._lsb(idx)
        return ans